libsolutil/CommonData.h

```diff
@@ -141,6 +142,36 @@ inline std::multiset<T...>& operator-=(std::multiset<T...>& _a, C const& _b)
 namespace solidity::util
 {
 
+/// Functional map.
+/// Returns a container _oc applying @param _op to each element in @param _c.
+/// By default _oc is a vector.
+/// If another return type is desired, an empty contained of that type
+/// is given as @param _oc.
+template<class Container, class Callable, class OutputContainer =
+	std::vector<std::invoke_result_t<
+		Callable,
+		decltype(*std::begin(std::declval<Container>()))
+>>>
+auto fmap(Container const& _c, Callable&& _op, OutputContainer _oc = OutputContainer{})
+{
+	std::transform(std::begin(_c), std::end(_c), std::back_inserter(_oc), _op);
+	return _oc;
+}
+
+/// Functional fold.
+/// Given a container @param _c, an initial value @param _acc,
+/// and a binary operator @param _binaryOp(T, U), accumulate
+/// the elements of _c over _acc.
+/// Note that <numeric> has a similar function `accumulate` which
+/// until C++20 does *not* std::move the partial accumulated.
+template<class C, class T, class Callable>
+auto fold(C const& _c, T _acc, Callable&& _binaryOp)
```

Comment:
Is it possible to use `T _ acc = T{}`?

Reply:
And `T = std::invoke_result_t<
		Callable,
		decltype(*std::begin(std::declval<C>()))`

Reply:
`T _acc = T{}` I don't think so, because `T` must be known in order to derive the types in `Callable`, otherwise we need to give the template types when using it.

Reply:
> And `T = std::invoke_result_t< Callable, decltype(*std::begin(std::declval<C>()))`

Same thing here. We're using `T` to infer `Callable`. We can also do it the other way around, but I think it would become more verbose on the caller side

Reply:
But can't `T` be derived from the return type of `Callable`?

Reply:
> But can't `T` be derived from the return type of `Callable`?

If you fully specify the lambda, yes